src/main/java/duke/Storage.java

```diff
@@ -0,0 +1,226 @@
+package duke;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.format.DateTimeParseException;
+import java.util.ArrayList;
+import java.util.List;
+
+import duke.exception.CalendarException;
+import duke.exception.DukeException;
+import duke.exception.StorageException;
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.Todo;
+
+public class Storage {
+
+    private static String filePath;
+    //private TaskList taskList;
+
+    public Storage(String filePath) {
+        Storage.filePath = filePath;
+    }
+
+    /**
+     * Reads the file from hard disk.
+     * @return List containing the parsed data from file.
+     * @throws IOException   If file cannot be read or found.
+     * @throws DukeException If there is incorrect format in the file.
+     */
+    public static List<Task> readFile() throws IOException, DukeException {
+        Path folder = Path.of("data");
+        Path file = folder.resolve(filePath);
+
+        //Create a new directory by creating all parent directories first
+        Files.createDirectories(folder);
+
+        //if user is new and file does not exist, create the file
+        if (!Files.exists(file)) {
+            Files.createFile(file);
+        }
+
+        //read from file
+        BufferedReader reader = Files.newBufferedReader(file);
+        List<Task> tasks = new ArrayList<>(); //this does not update this.todos
+        String currentLine;
+        while ((currentLine = reader.readLine()) != null) {
+            try {
+                System.out.println(currentLine);
+                Task task = parseData(currentLine);
+                tasks.add(task);
+            } catch (StorageException e) {
+                System.out.println(e.getMessage());
+            } catch (DukeException d) {
+                System.out.println(d.getMessage());
+            }
+        }
+        return tasks;
+    }
+
+    /**
+     * Formats the data to more readable format.
+     * @param line String to be parsed.
+     * @return parsed Task.
+     * @throws DukeException If there is incorrect data format.
+     */
+    private static Task parseData(String line) throws DukeException {
```

Comment:
Perhaps you could avoid deep nesting in the body of this method? Since this method is longer than 30 LOC, maybe you could split into several methods or classes?

Reply:
I am having this issue as well. Perhaps we should shorten out method and use more helpers ?

Reply:
Hello @ameliatjy and @Diwu-Yi ! Thank you for your feedback. I am aware of this and will try my best to abstract and shorten my code. Thank you you for pointing it out 😃